<a href="https://colab.research.google.com/github/DrKAI/Example/blob/master/02-8%20%EC%9E%90%EB%A3%8C%ED%98%95%EC%9D%98%20%EA%B0%92%EC%9D%84%20%EC%A0%80%EC%9E%A5%ED%95%98%EB%8A%94%20%EA%B3%B5%EA%B0%84,%20%EB%B3%80%EC%88%98.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 변수는 어떻게 만들까?

In [0]:
a = 1
b = "python"
c = [1, 2, 3]

우리는 이미 앞에서부터 변수를 사용해왔다. 위의 a, b, c를 변수라 한다.

변수를 만들 때, **= (assignment)** 기호를 사용한다. C나 Java처럼 변수의 자료형까지 함께 쓸 필요는 없다. 파이썬은 스스로 판단한다.

**변수명 = 변수에 저장할 값**

# 변수란?

파이썬에서 하용하는 변수는 객체를 가리키는 것이라고도 할 수 있다. 객체란 우리가 지금껏 보아왔던 자료형과 같은 것을 의미한다.

클래스 챕터에서 자세하게 공부합시다.

In [0]:
a = [1, 2, 3]

[1, 2, 3]이라는 값을 가지는 리스트 자료형(객체)이 자동으로 메모리에 생성되고, 변수 a는 [1, 2, 3] 리스트가 저장된 메모리의 주소를 가리키게 된다.

a 변수가 가리키는 메모리의 주소는 아래의 예시를 보라.

In [2]:
a = [1, 2, 3]

id(a)

140668482739080

id 함수는 변수가 가리키고 있는 객체의 주소를 리턴해주는 파이썬 내장함수이다.

변수 a가 가리키는 [1, 2, 3] 리스트의 주소는 140668482739080이다.

# 리스트를 변수에 넣고 복사하고자 할 때

"복사"는 혼동하기 쉽다!

In [3]:
a = [1, 2, 3]
b = a

print(id(a))
print(id(b))

140668482739464
140668482739464


변수 b에 변수 a를 대입해하였고, a와 b의 주소를 확인했다.

보다시피, a와 b는 완전히 동일하다. [1, 2, 3] 리스트를 참조하는 변수의 갯수가 a와 b, 2개가 되었을 뿐이다.

동일한 객체를 가리키는지 판별하는 파이썬 명령어 **is**를 사용해보자. True를 리턴할 것이다.

In [5]:
a = [1, 2, 3]
b = a

print(a is b)

True


In [6]:
a = [1, 2, 3, 4, 5]
b = a

a[2] = 100

print(a)
print(b)

[1, 2, 100, 4, 5]
[1, 2, 100, 4, 5]


a 리스트의 두번째 요소를 100으로 바꾸었더니 b 리스트 역시 바뀌었다. 상술하였듯, a와 b는 모두 동일한 리스트를 가리키고 있기 때문이다.

**그렇다면 변수 b를 생성할 때 변수 a와 같은 값을 가지도록 복사하면서 a가 가리키는 리스트와는 다른 리스트를 가리키게 할 방법은 뭐가 있을까?**

## 1. [:]

리스트 전체를 가리키는 [:]를 이용하여 복사한다.

In [10]:
a = [1, 2, 3, 4, 5]
b = a[:]
print(a)
print(b)

a[2] = 100
print(a)
print(b)

# 당연하게도 주소값도 달라진다!
print(id(a))
print(id(b))
print(a is b)

[1, 2, 3, 4, 5]
[1, 2, 3, 4, 5]
[1, 2, 100, 4, 5]
[1, 2, 3, 4, 5]
140668482738248
140668482739784
False


## 2. copy 모듈 이용

상세한 설명은 파이썬 모듈 파트에서 하겠다. **아래의 예 b = copy(a)는 b = a[:]과 같다!**

In [12]:
from copy import copy

a = [1, 2, 3, 4, 5]
b = copy(a)
print(a)
print(b)

a[2] = 65535
print(a)
print(b)


print(id(a))
print(id(b))
print(b is a)

[1, 2, 3, 4, 5]
[1, 2, 3, 4, 5]
[1, 2, 65535, 4, 5]
[1, 2, 3, 4, 5]
140668479928456
140668482738248
False


# 변수를 만드는 여러가지 방법

In [5]:
a, b = ('python', 'life')

print (a, b)

python life


In [8]:
(a, b) = 'Life is', 'unbeautiful'

print (a, b)

Life is unbeautiful


위의 두 예문은 완전히 동일한 방법이다. **튜플은 괄호를 생략해도 됩니다!**

아래처럼 리스트로도 변수를 만들 수 있다.

In [15]:
[a, b] = ['파이썬', '인생']

print (a, b)

파이썬 인생
<class 'str'>


여러 개의 변수에 같은 값을 대입할 수도 있다.

In [16]:
a = b = 'Hell'

print (a, b)

Hell Hell


두 변수의 값을 아래처럼 서로 간단하게 바꿀 수도 있다.

In [17]:
a = "Heaven"
b = "Hell"

print (a)
print (b)

a, b = b, a

print (a)
print (b)

Heaven
Hell
Hell
Heaven


# 연습문제

## Q1. 변수와 객체1

아래 예제를 실행하고 그 결과를 설명하라.

**a = [1. 2. 3]**

**b = [1. 2. 3]**

**a is b**

In [21]:
# 리스트 a와 b가 동일한 지 생각해보라는 거 같은데, 같은 리스트가 아닐 것이다.

a = [1, 2, 3]
b = [1, 2, 3]
print (a is b)

False


## Q2. 변수와 객체 2

아래 예제를 실행하고 그 결과를 설명하라.

**a = [1, 2, 3]**

**b = a**

**print (a is b)**

In [20]:
# 리스트 a와 b를 완전하게(주소값마저) 같게 만들려면 = 를 사용해야 한다.

a = [1, 2, 3]
b = a
print (a is b)

True


## Q3. 객체의 변경

파이썬은 동일한 값에 여러 개의 변수를 선언할 수 있다.

아래와 같이 a, b 변수를 선언한 후 a의 첫 번째 요소값을 변경하면 b의 값은 어떻게 될까?

그리고 그 이유를 설명해보라.

**a = b = [1, 2, 3]**

**a[1] = 100**

**print (b)**

In [23]:
# 리스트 b의 두 번째 요소값도 100으로 변할 것이다. 변수 a와 b는 같은 리스트를 지목하고 있으니.

a = b = [1, 2, 3]
a[1] = 100
print (b)

[1, 100, 3]


## Q4. 리스트 복사1

아래 예제를 실행하고 그 결과를 설명하라.

**a = [1, 2, 3]**

**b = a[:]**

**a is b**

In [24]:
# 이건 서로 다르게 나올 것이다. 복사는 주소가 서로 다른 곳에 내용물만 같게 만드는 것이다.

a = [1, 2, 3]
b = a[:]
print (a is b)

False


## Q5. 리스트 복사2

리스트 b는 리스트 a를 copy하여 다음과 같이 생성하였다.

**a = [1, 2, 3]**

**b = a[:]**

그리고 다음과 같이 리스트 a의 두 번째 요소값을 2에서 10000으로 바꾸었다.

**a[1] = 10000**

**a**

**print (b)**

이제 리스트 b를 출력하면 어떤 값이 나올까? 그리고 그 결과값이 나오는 이유를 설명하라.

In [25]:
# 이전 문제에서 상술하였듯, 주소값이 아예 달라진다. 리스트 b의 두 번째 요소값은 여전히 2일 것이다.
# 그렇기에 변수 a와 b는 내용물은 같지만, 완전히 다른 리스트를 지목하고 있다.

a = [1, 2, 3]
b = a[:]
a[1] = 10000

print (a)
print (b)

[1, 10000, 3]
[1, 2, 3]


## Q6. 리스트의 더하기와 extend

+ 기호를 이용하여 더한 것과 extend한 것의 차이점이 있을까?

In [26]:
a = [1, 2, 3]
a = a + [10, 100, 10000]

print (a)

[1, 2, 3, 10, 100, 10000]


In [27]:
a = [1, 2, 3]
a.extend([10, 100, 10000])

print (a)

[1, 2, 3, 10, 100, 10000]


In [42]:
# 덧셈 기호(+) 테스트 : 잘못된 테스트. 변수를 새로 받아하는 방식은 잘못 됨

a = [1, 2, 3]
b = a
print (a)
print (b)

a = a + [10, 100, 10000]
print (a)
print (b)

print (a is b)

[1, 2, 3]
[1, 2, 3]
[1, 2, 3, 10, 100, 10000]
[1, 2, 3]
False


In [43]:
# extend 테스트 : 잘못된 테스트. 변수를 새로 받아하는 방식은 잘못 됨

a = [1, 2, 3]
b = a
print (a)
print (b)

a.extend([10, 100, 10000])
print (a)
print (b)

print (a is b)

[1, 2, 3]
[1, 2, 3]
[1, 2, 3, 10, 100, 10000]
[1, 2, 3, 10, 100, 10000]
True


In [47]:
# 주소값 비교 : 덧셈 기호 (+)

a = [1, 2, 3]
print (a)
print (id(a))

a = a + [10, 100, 10000]
print (a)
print (id(a))

[1, 2, 3]
140716647300104
[1, 2, 3, 10, 100, 10000]
140716647421512


In [46]:
# 주소값 비교 : extend

a = [1, 2, 3]
print (a)
print (id(a))

a.extend([10, 100, 10000])
print (a)
print (id(a))

[1, 2, 3]
140716647423560
[1, 2, 3, 10, 100, 10000]
140716647423560


**덧셈 기호(+)를 통해 리스트 추가 시 주소값이 달라진다!**

**extend는 주소값이 유지된다!**

## Q7. 리스트 복사3

**a = [1, [2, 3], 4]**

**b = a[:]**

**a = [1][0] = 10000**

**print (b)**

리스트 b는 어떻게 될까? 그 이유도 설명하라.

In [54]:
# 아무래도 리스트 b 역시 요소값이 변경할 것 같다.
# a[1][0]의 의미는 리스트 a의 두번째 요소값에 해당하는 리스트의 첫번째 요소값을 변경한다는 의미이다.

a = [1, [2, 3], 4]
b = a[:]
print (a)
print (b)

a[1][0] = 10000
print (a)
print (b)

print (a is b)

[1, [2, 3], 4]
[1, [2, 3], 4]
[1, [10000, 3], 4]
[1, [10000, 3], 4]
False


**그렇지만 주소값마저 복사하는 것은 아니기에 변수 a와 b는 주소값이 서로 다르다.**

**from import로 deep copy를 하면, 리스트 a의 주소값 자체를 복사하기에 리스트 b는 변경 이전 a 리스트의 형태를 했을 것이다.**